ideascube/serveradmin/catalog.py

```diff
@@ -0,0 +1,459 @@
+
+            with open(libpath, 'r') as f:
+                et = etree.parse(f)
+                books = et.findall('book')
+
+                # Messing with the path gets complicated otherwise
+                # TODO: Can we assume this is always true for stuff distributed
+                #       by kiwix?
+                assert len(books) == 1
+
+                book = books[0]
+                book.set('path', 'data/content/%s' % zimname)
+                book.set('indexPath', 'data/content/%s.idx' % zimname)
+
+                library.append(book)
+
+        with open(self._library, 'wb') as f:
+            f.write(etree.tostring(
+                library, xml_declaration=True, encoding='utf-8'))
+
+
+# Register the handler types
+Handler.__TYPES__ = {
+    cls.__name__.lower(): cls for cls in Handler.__subclasses__()
+    }
+
+
+class Catalog:
+    def __init__(self, cache_base_dir):
+        self._cache_base_dir = cache_base_dir
+        os.makedirs(cache_base_dir, exist_ok=True)
+
+        self._cache_remote_dir = os.path.join(cache_base_dir, 'remotes')
+        self._load_remotes()
+
+        self._cache_catalog = os.path.join(cache_base_dir, 'catalog.yml')
+        self._local_package_cache = os.path.join(cache_base_dir, 'packages')
+        self._load_cache()
+
+    def _progress(self, msg, i, chunk_size, remote_size):
+        percent_done = (i + 1) * chunk_size / remote_size
+
+        if percent_done > 1.0:
+            percent_done = 1.0
+
+        length = 79 - len(msg) - 2 - 1 - 6
+        done_chars = int(percent_done * length)
+        remain_chars = length - done_chars
+        percent_done = int(percent_done * 1000) / 10
+
+        p = '\r{}: {}{} {}%'.format(
+            msg, "█" * done_chars, " " * remain_chars, percent_done)
+        sys.stdout.write(p)
+
+        if percent_done == 100.0:
+            sys.stdout.write('\n')
+
+        sys.stdout.flush()
+
+    # -- Manage packages ------------------------------------------------------
+    def _get_package(self, id, source):
+        try:
+            metadata = source[id]
+
+        except KeyError:
+            raise NoSuchPackage(id)
+
+        return Package(id, metadata)
+
+    def _get_handler(self, package):
+        return Handler(package.handler)
+
+    def _verify_sha256(self, path, sha256sum):
+        sha = sha256()
+
+        with open(path, 'rb') as f:
+            while True:
+                data = f.read(8388608)
+
+                if not data:
+                    break
+
+                sha.update(data)
+
+        return sha.hexdigest() == sha256sum
+
+    def fetch_package(self, package):
+        def _progress(*args):
+            self._progress(' {}'.format(package.id), *args)
+
+        filename = '{0.id}-{0.version}'.format(package)
+
+        for cache in self._package_caches:
+            path = os.path.join(cache, filename)
+
+            if os.path.isfile(path):
+                if self._verify_sha256(path, package.sha256sum):
```

Comment:
Not verifying checksums seems like a giant security hole. How do you ensure you're installing the right content without checking the checksum?

To me, we should at the minimum:

* download **everything** over https
* verify the `catalog.yml` downloaded from the remote source
* verify the downloaded packages with a strong checksum (afaik sha256sum is still ok these days)

Currently we're not doing the first two, so the third one is not very useful. But that seems like an argument for doing the first two, rather than an argument for removing the third one. :smile: